tt_metal/llrt/llrt.cpp

```diff
@@ -24,44 +24,31 @@ using std::uint16_t;
 using std::uint32_t;
 using std::uint64_t;
 
-struct HexNameToMemVectorCache {
-    using lock = std::unique_lock<std::mutex>;
-    // maps from RisckCacheMapKey to hex file path
-    static HexNameToMemVectorCache &inst() {
-        static HexNameToMemVectorCache inst_;
-        return inst_;
-    }
-
-    bool exists(const string &path) {
-        lock l(mutex_);
-        return cache_.find(path) != cache_.end();
-    }
-    ll_api::memory &get(const string &path) {
-        lock l(mutex_);
-        return cache_[path];
-    }
-    void add(const string &path, ll_api::memory &mem) {
-        lock l(mutex_);
-        cache_[path] = mem;
-    }
-
-    std::unordered_map<std::string, ll_api::memory> cache_;
-    std::mutex mutex_;
-};
-
 ll_api::memory get_risc_binary(std::string path) {
-
-    if (HexNameToMemVectorCache::inst().exists(path)) {
-        return HexNameToMemVectorCache::inst().get(path);
-    }
-
+  static struct {
+    std::unordered_map<std::string, std::unique_ptr<ll_api::memory>> map;
+    std::mutex mutex;
+    std::condition_variable cvar;
+  } cache;
+
+  std::unique_lock lock(cache.mutex);
+  auto [slot, inserted] = cache.map.try_emplace(path);
+  if (inserted) {
+    // We're the first with PATH. Create and insert.
+    lock.unlock();
     std::ifstream hex_istream(path);
-    ll_api::memory mem(hex_istream);
-
-    // add this path to binary cache
-    HexNameToMemVectorCache::inst().add(path, mem);
-
-    return mem;
+    auto *ptr = new ll_api::memory(hex_istream);
+    lock.lock();
+    // maps have iterator stability, so SLOT is still valid.
+    slot->second = decltype(slot->second)(ptr);
+    // We can't wake just those waiting on this slot, so wake them
+    // all. Should be a rare event anyway.
+    cache.cvar.notify_all();
+  } else if (!slot->second)
+    // Someone else is creating the initial entry, wait for them.
+    cache.cvar.wait(lock, [=] { return bool(slot->second); });
```

Comment:
forgive my ignorance, but if !inserted and !second, we call wait on the lock but if the lock was released in the middle of the conditional above, won't wait return immediately (ie, lock is available) w/ second still uniniitalized?

Reply:
TLDR: No.  Long version: cvars (and mutexes) are designed for exactly this use. We hold the lock until we get to the wait, and then atomically unlock and queue on the cvar. The other thread that's doing the creation will get the lock after creation, install the object the signal the waiters and then unlock. The woken waiters will queue getting the mutex back, which only happens once the signaller has unlocked it. Does that help?
https://en.cppreference.com/w/cpp/thread/condition_variable